In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
% matplotlib inline
import matplotlib as plt
import csv
import os
import requests
import tweepy

### Gather: Twitter Archive Enhanced###
**Let's start with gathering the enhanced twitter archive data**

In [2]:
df = pd.read_csv('twitter-archive-enhanced.csv')

In [3]:
df.head()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
0  892420643555336193                    NaN                  NaN   
1  892177421306343426                    NaN                  NaN   
2  891815181378084864                    NaN                  NaN   
3  891689557279858688                    NaN                  NaN   
4  891327558926688256                    NaN                  NaN   

                   timestamp  \
0  2017-08-01 16:23:56 +0000   
1  2017-08-01 00:17:27 +0000   
2  2017-07-31 00:18:03 +0000   
3  2017-07-30 15:58:51 +0000   
4  2017-07-29 16:00:24 +0000   

                                              source  \
0  <a href="http://twitter.com/download/iphone" r...   
1  <a href="http://twitter.com/download/iphone" r...   
2  <a href="http://twitter.com/download/iphone" r...   
3  <a href="http://twitter.com/download/iphone" r...   
4  <a href="http://twitter.com/download/iphone" r...   

                                                text  retweeted_status_id  \
0  This is Phineas. He's a mystical boy. Only eve...                  NaN   
1  This is Tilly. She's just checking pup on you....                  NaN   
2  This is Archie. He is a rare Norwegian Pouncin...                  NaN   
3  This is Darla. She commenced a snooze mid meal...                  NaN   
4  This is Franklin. He would like you to stop ca...                  NaN   

   retweeted_status_user_id retweeted_status_timestamp  \
0                       NaN                        NaN   
1                       NaN                        NaN   
2                       NaN                        NaN   
3                       NaN                        NaN   
4                       NaN                        NaN   

                                       expanded_urls  rating_numerator  \
0  https://twitter.com/dog_rates/status/892420643...                13   
1  https://twitter.com/dog_rates/status/892177421...                13   
2  https://twitter.com/dog_rates/status/891815181...                12   
3  https://twitter.com/dog_rates/status/891689557...                13   
4  https://twitter.com/dog_rates/status/891327558...                12   

   rating_denominator      name doggo floofer pupper puppo  
0                  10   Phineas  None    None   None  None  
1                  10     Tilly  None    None   None  None  
2                  10    Archie  None    None   None  None  
3                  10     Darla  None    None   None  None  
4                  10  Franklin  None    None   None  None

In [4]:
df.tail()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
2351  666049248165822465                    NaN                  NaN   
2352  666044226329800704                    NaN                  NaN   
2353  666033412701032449                    NaN                  NaN   
2354  666029285002620928                    NaN                  NaN   
2355  666020888022790149                    NaN                  NaN   

                      timestamp  \
2351  2015-11-16 00:24:50 +0000   
2352  2015-11-16 00:04:52 +0000   
2353  2015-11-15 23:21:54 +0000   
2354  2015-11-15 23:05:30 +0000   
2355  2015-11-15 22:32:08 +0000   

                                                 source  \
2351  <a href="http://twitter.com/download/iphone" r...   
2352  <a href="http://twitter.com/download/iphone" r...   
2353  <a href="http://twitter.com/download/iphone" r...   
2354  <a href="http://twitter.com/download/iphone" r...   
2355  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
2351  Here we have a 1949 1st generation vulpix. Enj...                  NaN   
2352  This is a purebred Piers Morgan. Loves to Netf...                  NaN   
2353  Here is a very happy pup. Big fan of well-main...                  NaN   
2354  This is a western brown Mitsubishi terrier. Up...                  NaN   
2355  Here we have a Japanese Irish Setter. Lost eye...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
2351                       NaN                        NaN   
2352                       NaN                        NaN   
2353                       NaN                        NaN   
2354                       NaN                        NaN   
2355                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
2351  https://twitter.com/dog_rates/status/666049248...                 5   
2352  https://twitter.com/dog_rates/status/666044226...                 6   
2353  https://twitter.com/dog_rates/status/666033412...                 9   
2354  https://twitter.com/dog_rates/status/666029285...                 7   
2355  https://twitter.com/dog_rates/status/666020888...                 8   

      rating_denominator  name doggo floofer pupper puppo  
2351                  10  None  None    None   None  None  
2352                  10     a  None    None   None  None  
2353                  10     a  None    None   None  None  
2354                  10     a  None    None   None  None  
2355                  10  None  None    None   None  None

In [5]:
df.query('name == "a" or name == "an"').name.count()

62

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
tweet_id                      2356 non-null int64
in_reply_to_status_id         78 non-null float64
in_reply_to_user_id           78 non-null float64
timestamp                     2356 non-null object
source                        2356 non-null object
text                          2356 non-null object
retweeted_status_id           181 non-null float64
retweeted_status_user_id      181 non-null float64
retweeted_status_timestamp    181 non-null object
expanded_urls                 2297 non-null object
rating_numerator              2356 non-null int64
rating_denominator            2356 non-null int64
name                          2356 non-null object
doggo                         2356 non-null object
floofer                       2356 non-null object
pupper                        2356 non-null object
puppo                         2356 non-null object
dtypes: float64(4), int64(3), ob

In [7]:
df.describe()

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
count  2.356000e+03           7.800000e+01         7.800000e+01   
mean   7.427716e+17           7.455079e+17         2.014171e+16   
std    6.856705e+16           7.582492e+16         1.252797e+17   
min    6.660209e+17           6.658147e+17         1.185634e+07   
25%    6.783989e+17           6.757419e+17         3.086374e+08   
50%    7.196279e+17           7.038708e+17         4.196984e+09   
75%    7.993373e+17           8.257804e+17         4.196984e+09   
max    8.924206e+17           8.862664e+17         8.405479e+17   

       retweeted_status_id  retweeted_status_user_id  rating_numerator  \
count         1.810000e+02              1.810000e+02       2356.000000   
mean          7.720400e+17              1.241698e+16         13.126486   
std           6.236928e+16              9.599254e+16         45.876648   
min           6.661041e+17              7.832140e+05          0.000000   
25%           7.186315e+17              4.196984e+09         10.000000   
50%           7.804657e+17              4.196984e+09         11.000000   
75%           8.203146e+17              4.196984e+09         12.000000   
max           8.874740e+17              7.874618e+17       1776.000000   

       rating_denominator  
count         2356.000000  
mean            10.455433  
std              6.745237  
min              0.000000  
25%             10.000000  
50%             10.000000  
75%             10.000000  
max            170.000000

In [8]:
df[df.rating_numerator > 20].tweet_id.count()

24

In [9]:
#some ratings are extremely high (max is 2356) and some rating denominators are less than 10

In [10]:
df[df.rating_denominator != 10].tweet_id.count()

23

In [11]:
df[df.in_reply_to_status_id.notnull()]
#78 instances of reply tweets for which we want to ignore

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
30    886267009285017600           8.862664e+17         2.281182e+09   
55    881633300179243008           8.816070e+17         4.738443e+07   
64    879674319642796034           8.795538e+17         3.105441e+09   
113   870726314365509632           8.707262e+17         1.648776e+07   
148   863427515083354112           8.634256e+17         7.759620e+07   
149   863079547188785154           6.671522e+17         4.196984e+09   
179   857214891891077121           8.571567e+17         1.806710e+08   
184   856526610513747968           8.558181e+17         4.196984e+09   
186   856288084350160898           8.562860e+17         2.792810e+08   
188   855862651834028034           8.558616e+17         1.943518e+08   
189   855860136149123072           8.558585e+17         1.361572e+07   
218   850333567704068097           8.503288e+17         2.195506e+07   
228   848213670039564288           8.482121e+17         4.196984e+09   
234   847617282490613760           8.476062e+17         4.196984e+09   
251   844979544864018432           7.590995e+17         4.196984e+09   
274   840698636975636481           8.406983e+17         8.405479e+17   
290   838150277551247360           8.381455e+17         2.195506e+07   
291   838085839343206401           8.380855e+17         2.894131e+09   
313   835246439529840640           8.352460e+17         2.625958e+07   
342   832088576586297345           8.320875e+17         3.058208e+07   
346   831926988323639298           8.319030e+17         2.068372e+07   
387   826598799820865537           8.265984e+17         4.196984e+09   
409   823333489516937216           8.233264e+17         1.582854e+09   
427   821153421864615936           8.211526e+17         1.132119e+08   
498   813130366689148928           8.131273e+17         4.196984e+09   
513   811647686436880384           8.116272e+17         4.196984e+09   
565   802265048156610565           7.331095e+17         4.196984e+09   
570   801854953262350336           8.018543e+17         1.185634e+07   
576   800859414831898624           8.008580e+17         2.918590e+08   
611   797165961484890113           7.971238e+17         2.916630e+07   
...                  ...                    ...                  ...   
1479  693582294167244802           6.935722e+17         1.198989e+09   
1497  692423280028966913           6.924173e+17         4.196984e+09   
1501  692142790915014657           6.920419e+17         4.196984e+09   
1523  690607260360429569           6.903413e+17         4.670367e+08   
1598  686035780142297088           6.860340e+17         4.196984e+09   
1605  685681090388975616           6.855479e+17         4.196984e+09   
1618  684969860808454144           6.849598e+17         4.196984e+09   
1630  684538444857667585           6.844811e+17         4.196984e+09   
1634  684225744407494656           6.842229e+17         4.196984e+09   
1663  682808988178739200           6.827884e+17         4.196984e+09   
1689  681340665377193984           6.813394e+17         4.196984e+09   
1774  678023323247357953           6.780211e+17         4.196984e+09   
1819  676590572941893632           6.765883e+17         4.196984e+09   
1842  675870721063669760           6.757073e+17         4.196984e+09   
1844  675849018447167488           6.758457e+17         4.196984e+09   
1852  675707330206547968           6.754971e+17         4.196984e+09   
1866  675349384339542016           6.749998e+17         4.196984e+09   
1882  674999807681908736           6.747934e+17         4.196984e+09   
1885  674793399141146624           6.717299e+17         4.196984e+09   
1892  674754018082705410           6.747522e+17         4.196984e+09   
1895  674742531037511680           6.747400e+17         4.196984e+09   
1905  674606911342424069           6.744689e+17         4.196984e+09   
1914  674330906434379776           6.658147e+17         1.637468e+07   
1940  673716320723169284           6.737159e+17         

In [12]:
df[(df.in_reply_to_status_id.notnull()) & (df.name == 'None')]
#77 instances where such replies have no dog names -- so confirming these are not original tweets and can be deleted

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
30    886267009285017600           8.862664e+17         2.281182e+09   
55    881633300179243008           8.816070e+17         4.738443e+07   
64    879674319642796034           8.795538e+17         3.105441e+09   
113   870726314365509632           8.707262e+17         1.648776e+07   
148   863427515083354112           8.634256e+17         7.759620e+07   
149   863079547188785154           6.671522e+17         4.196984e+09   
179   857214891891077121           8.571567e+17         1.806710e+08   
184   856526610513747968           8.558181e+17         4.196984e+09   
186   856288084350160898           8.562860e+17         2.792810e+08   
188   855862651834028034           8.558616e+17         1.943518e+08   
189   855860136149123072           8.558585e+17         1.361572e+07   
218   850333567704068097           8.503288e+17         2.195506e+07   
228   848213670039564288           8.482121e+17         4.196984e+09   
234   847617282490613760           8.476062e+17         4.196984e+09   
251   844979544864018432           7.590995e+17         4.196984e+09   
274   840698636975636481           8.406983e+17         8.405479e+17   
290   838150277551247360           8.381455e+17         2.195506e+07   
291   838085839343206401           8.380855e+17         2.894131e+09   
313   835246439529840640           8.352460e+17         2.625958e+07   
342   832088576586297345           8.320875e+17         3.058208e+07   
346   831926988323639298           8.319030e+17         2.068372e+07   
387   826598799820865537           8.265984e+17         4.196984e+09   
409   823333489516937216           8.233264e+17         1.582854e+09   
427   821153421864615936           8.211526e+17         1.132119e+08   
498   813130366689148928           8.131273e+17         4.196984e+09   
513   811647686436880384           8.116272e+17         4.196984e+09   
565   802265048156610565           7.331095e+17         4.196984e+09   
570   801854953262350336           8.018543e+17         1.185634e+07   
576   800859414831898624           8.008580e+17         2.918590e+08   
611   797165961484890113           7.971238e+17         2.916630e+07   
...                  ...                    ...                  ...   
1474  693644216740769793           6.936422e+17         4.196984e+09   
1479  693582294167244802           6.935722e+17         1.198989e+09   
1497  692423280028966913           6.924173e+17         4.196984e+09   
1501  692142790915014657           6.920419e+17         4.196984e+09   
1523  690607260360429569           6.903413e+17         4.670367e+08   
1598  686035780142297088           6.860340e+17         4.196984e+09   
1605  685681090388975616           6.855479e+17         4.196984e+09   
1618  684969860808454144           6.849598e+17         4.196984e+09   
1630  684538444857667585           6.844811e+17         4.196984e+09   
1634  684225744407494656           6.842229e+17         4.196984e+09   
1663  682808988178739200           6.827884e+17         4.196984e+09   
1689  681340665377193984           6.813394e+17         4.196984e+09   
1774  678023323247357953           6.780211e+17         4.196984e+09   
1819  676590572941893632           6.765883e+17         4.196984e+09   
1842  675870721063669760           6.757073e+17         4.196984e+09   
1844  675849018447167488           6.758457e+17         4.196984e+09   
1852  675707330206547968           6.754971e+17         4.196984e+09   
1866  675349384339542016           6.749998e+17         4.196984e+09   
1882  674999807681908736           6.747934e+17         4.196984e+09   
1885  674793399141146624           6.717299e+17         4.196984e+09   
1892  674754018082705410           6.747522e+17         4.196984e+09   
1895  674742531037511680           6.747400e+17         4.196984e+09   
1905  674606911342424069           6.744689e+17         4.196984e+09   
1914  674330906434379776           6.658147e+17         

In [13]:
df[(df.rating_numerator > 20) & (df.in_reply_to_status_id.isnull())]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
340   832215909146226688                    NaN                  NaN   
433   820690176645140481                    NaN                  NaN   
516   810984652412424192                    NaN                  NaN   
695   786709082849828864                    NaN                  NaN   
763   778027034220126208                    NaN                  NaN   
902   758467244762497024                    NaN                  NaN   
979   749981277374128128                    NaN                  NaN   
1120  731156023742988288                    NaN                  NaN   
1202  716439118184652801                    NaN                  NaN   
1228  713900603437621249                    NaN                  NaN   
1254  710658690886586372                    NaN                  NaN   
1274  709198395643068416                    NaN                  NaN   
1351  704054845121142784                    NaN                  NaN   
1433  697463031882764288                    NaN                  NaN   
1635  684222868335505415                    NaN                  NaN   
1712  680494726643068929                    NaN                  NaN   
1779  677716515794329600                    NaN                  NaN   
1843  675853064436391936                    NaN                  NaN   
2074  670842764863651840                    NaN                  NaN   

                      timestamp  \
340   2017-02-16 13:11:49 +0000   
433   2017-01-15 17:52:40 +0000   
516   2016-12-19 23:06:23 +0000   
695   2016-10-13 23:23:56 +0000   
763   2016-09-20 00:24:34 +0000   
902   2016-07-28 01:00:57 +0000   
979   2016-07-04 15:00:45 +0000   
1120  2016-05-13 16:15:54 +0000   
1202  2016-04-03 01:36:11 +0000   
1228  2016-03-27 01:29:02 +0000   
1254  2016-03-18 02:46:49 +0000   
1274  2016-03-14 02:04:08 +0000   
1351  2016-02-28 21:25:30 +0000   
1433  2016-02-10 16:51:59 +0000   
1635  2016-01-05 04:00:18 +0000   
1712  2015-12-25 21:06:00 +0000   
1779  2015-12-18 05:06:23 +0000   
1843  2015-12-13 01:41:41 +0000   
2074  2015-11-29 05:52:33 +0000   

                                                 source  \
340   <a href="http://twitter.com/download/iphone" r...   
433   <a href="http://twitter.com/download/iphone" r...   
516   <a href="http://twitter.com/download/iphone" r...   
695   <a href="http://twitter.com/download/iphone" r...   
763   <a href="http://twitter.com/download/iphone" r...   
902   <a href="http://twitter.com/download/iphone" r...   
979   <a href="https://about.twitter.com/products/tw...   
1120  <a href="http://twitter.com/download/iphone" r...   
1202  <a href="http://twitter.com/download/iphone" r...   
1228  <a href="http://twitter.com/download/iphone" r...   
1254  <a href="http://twitter.com/download/iphone" r...   
1274  <a href="http://twitter.com/download/iphone" r...   
1351  <a href="http://twitter.com/download/iphone" r...   
1433  <a href="http://twitter.com/download/iphone" r...   
1635  <a href="http://twitter.com/download/iphone" r...   
1712  <a href="http://twitter.com/download/iphone" r...   
1779  <a href="http://twitter.com/download/iphone" r...   
1843  <a href="http://twitter.com/download/iphone" r...   
2074  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
340   RT @dog_rates: This is Logan, the Chow who liv...         7.867091e+17   
433   The floofs have been released I repeat the flo...                  NaN   
516   Meet Sam. She smiles 24/7 &amp; secretly aspir...                  NaN   
695   This is Logan, the Chow who lived. He solemnly...                  NaN   
763   This is Sophie. She's a Jubilant Bush Pupper. ...                  NaN   
902   Why does this never happen at my front door......                  NaN   
979   This is Atticus. He's quite simply America af....                  NaN   
1120  Say hello to this unbelievably well behaved sq.

In [14]:
df[(df.in_reply_to_status_id.notnull()) & (df.name != 'None')]
# we can keep this instance (669353438988365824) along with following tweet ids: 863079547188785154, 856526610513747968, 802265048156610565,
#675870721063669760, 675349384339542016, 674793399141146624, 671729906628341761

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
2169  669353438988365824           6.678065e+17         4.196984e+09   

                      timestamp  \
2169  2015-11-25 03:14:30 +0000   

                                                 source  \
2169  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
2169  This is Tessa. She is also very pleased after ...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
2169                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
2169  https://twitter.com/dog_rates/status/669353438...                10   

      rating_denominator   name doggo floofer pupper puppo  
2169                  10  Tessa  None    None   None  None

In [15]:
df[df.expanded_urls.isnull()]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
30    886267009285017600           8.862664e+17         2.281182e+09   
55    881633300179243008           8.816070e+17         4.738443e+07   
64    879674319642796034           8.795538e+17         3.105441e+09   
113   870726314365509632           8.707262e+17         1.648776e+07   
148   863427515083354112           8.634256e+17         7.759620e+07   
179   857214891891077121           8.571567e+17         1.806710e+08   
185   856330835276025856                    NaN                  NaN   
186   856288084350160898           8.562860e+17         2.792810e+08   
188   855862651834028034           8.558616e+17         1.943518e+08   
189   855860136149123072           8.558585e+17         1.361572e+07   
218   850333567704068097           8.503288e+17         2.195506e+07   
228   848213670039564288           8.482121e+17         4.196984e+09   
234   847617282490613760           8.476062e+17         4.196984e+09   
274   840698636975636481           8.406983e+17         8.405479e+17   
290   838150277551247360           8.381455e+17         2.195506e+07   
291   838085839343206401           8.380855e+17         2.894131e+09   
313   835246439529840640           8.352460e+17         2.625958e+07   
342   832088576586297345           8.320875e+17         3.058208e+07   
346   831926988323639298           8.319030e+17         2.068372e+07   
375   828361771580813312                    NaN                  NaN   
387   826598799820865537           8.265984e+17         4.196984e+09   
409   823333489516937216           8.233264e+17         1.582854e+09   
427   821153421864615936           8.211526e+17         1.132119e+08   
498   813130366689148928           8.131273e+17         4.196984e+09   
513   811647686436880384           8.116272e+17         4.196984e+09   
570   801854953262350336           8.018543e+17         1.185634e+07   
576   800859414831898624           8.008580e+17         2.918590e+08   
611   797165961484890113           7.971238e+17         2.916630e+07   
701   786051337297522688           7.727430e+17         7.305050e+17   
707   785515384317313025                    NaN                  NaN   
843   766714921925144576           7.667118e+17         4.196984e+09   
857   763956972077010945           7.638652e+17         1.584641e+07   
967   750381685133418496           7.501805e+17         4.717297e+09   
1005  747651430853525504           7.476487e+17         4.196984e+09   
1080  738891149612572673           7.384119e+17         3.589728e+08   
1295  707983188426153984           7.079801e+17         2.319108e+09   
1345  704491224099647488           7.044857e+17         2.878549e+07   
1445  696518437233913856                    NaN                  NaN   
1446  696490539101908992           6.964887e+17         4.196984e+09   
1474  693644216740769793           6.936422e+17         4.196984e+09   
1479  693582294167244802           6.935722e+17         1.198989e+09   
1497  692423280028966913           6.924173e+17         4.196984e+09   
1523  690607260360429569           6.903413e+17         4.670367e+08   
1598  686035780142297088           6.860340e+17         4.196984e+09   
1605  685681090388975616           6.855479e+17         4.196984e+09   
1618  684969860808454144           6.849598e+17         4.196984e+09   
1663  682808988178739200           6.827884e+17         4.196984e+09   
1689  681340665377193984           6.813394e+17         4.196984e+09   
1774  678023323247357953           6.780211e+17         4.196984e+09   
1819  676590572941893632           6.765883e+17         4.196984e+09   
1844  675849018447167488           6.758457e+17         4.196984e+09   
1895  674742531037511680           6.747400e+17         4.196984e+09   
1905  674606911342424069           6.744689e+17         4.196984e+09   
1914  674330906434379776           6.658147e+17         1.637468e+07   
1940  673716320723169284           6.737159e+17         

In [16]:
df.rating_numerator.isnull().any()
#no missing ratings

False

In [17]:
df.tweet_id.duplicated().any()
#no duplicated tweets

False

In [18]:
df.expanded_urls.duplicated().any()

True

In [19]:
df_url = df[df.expanded_urls.notna()]
#df.dropna().loc[df['foo'].duplicated()]
df_url[df_url.expanded_urls.duplicated(keep=False)].sort_values(by=['expanded_urls', 'text'])


tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
349   831650051525054464                    NaN                  NaN   
204   852936405516943360                    NaN                  NaN   
2338  666104133288665088                    NaN                  NaN   
605   798576900688019456                    NaN                  NaN   
2294  667138269671505920                    NaN                  NaN   
1043  743835915802583040                    NaN                  NaN   
273   840728873075638272                    NaN                  NaN   
2293  667152164079423490                    NaN                  NaN   
604   798585098161549313                    NaN                  NaN   
2286  667182792070062081                    NaN                  NaN   
1023  746521445350707200                    NaN                  NaN   
2248  667866724293877760                    NaN                  NaN   
2180  669000397445533696                    NaN                  NaN   
74    878316110768087041                    NaN                  NaN   
1970  673295268553605120                    NaN                  NaN   
860   763167063695355904                    NaN                  NaN   
1806  676936541936185344                    NaN                  NaN   
599   798682547630837760                    NaN                  NaN   
1742  679462823135686656                    NaN                  NaN   
749   779124354206535695                    NaN                  NaN   
1726  680055455951884288                    NaN                  NaN   
506   812747805718642688                    NaN                  NaN   
937   753298634498793472                    NaN                  NaN   
1688  681523177663676416                    NaN                  NaN   
583   800188575492947969                    NaN                  NaN   
1683  681694085539872773                    NaN                  NaN   
598   798686750113755136                    NaN                  NaN   
1655  683391852557561860                    NaN                  NaN   
612   796904159865868288                    NaN                  NaN   
1578  687317306314240000                    NaN                  NaN   
...                  ...                    ...                  ...   
91    874434818259525634                    NaN                  NaN   
136   866334964761202691                    NaN                  NaN   
132   866816280283807744                    NaN                  NaN   
135   866450705531457537                    NaN                  NaN   
68    879130579576475649                    NaN                  NaN   
76    878057613040115712                    NaN                  NaN   
2212  668587383441514497                    NaN                  NaN   
657   791774931465953280                    NaN                  NaN   
272   840761248237133825                    NaN                  NaN   
276   840632337062862849                    NaN                  NaN   
124   868639477480148993                    NaN                  NaN   
126   868552278524837888                    NaN                  NaN   
75    878281511006478336                    NaN                  NaN   
73    878404777348136964                    NaN                  NaN   
161   860563773140209665                    NaN                  NaN   
159   860981674716409858                    NaN                  NaN   
97    873337748698140672                    NaN                  NaN   
98    873213775632977920                    NaN                  NaN   
146   863471782782697472                    NaN                  NaN   
150   863062471531167744                    NaN                  NaN   
312   835264098648616962                    NaN                  NaN   
309   835536468978302976                    NaN                  NaN   
397   825120256414846976                    NaN                  NaN   
398   825026590719483904                    NaN         

In [20]:
#df[df.expanded_urls.duplicated()].tweet_id.count()
df_url[df_url.expanded_urls.duplicated()].tweet_id.count()

#there appear to be 79 such duplicate records

79

In [21]:
df[df.retweeted_status_id.notna()].tweet_id.count()
# there appear to be 181 retweets

181

In [22]:
df_url[df_url.retweeted_status_id.notna()].tweet_id.count()

180

In [23]:
df_url[df_url.retweeted_status_id.notna()].expanded_urls.duplicated().count()

180

In [24]:
df[df.retweeted_status_id.notna()]


tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
19    888202515573088257                    NaN                  NaN   
32    886054160059072513                    NaN                  NaN   
36    885311592912609280                    NaN                  NaN   
68    879130579576475649                    NaN                  NaN   
73    878404777348136964                    NaN                  NaN   
74    878316110768087041                    NaN                  NaN   
78    877611172832227328                    NaN                  NaN   
91    874434818259525634                    NaN                  NaN   
95    873697596434513921                    NaN                  NaN   
97    873337748698140672                    NaN                  NaN   
101   872668790621863937                    NaN                  NaN   
109   871166179821445120                    NaN                  NaN   
118   869988702071779329                    NaN                  NaN   
124   868639477480148993                    NaN                  NaN   
130   867072653475098625                    NaN                  NaN   
132   866816280283807744                    NaN                  NaN   
137   866094527597207552                    NaN                  NaN   
146   863471782782697472                    NaN                  NaN   
155   861769973181624320                    NaN                  NaN   
159   860981674716409858                    NaN                  NaN   
160   860924035999428608                    NaN                  NaN   
165   860177593139703809                    NaN                  NaN   
171   858860390427611136                    NaN                  NaN   
180   857062103051644929                    NaN                  NaN   
182   856602993587888130                    NaN                  NaN   
185   856330835276025856                    NaN                  NaN   
194   855245323840757760                    NaN                  NaN   
195   855138241867124737                    NaN                  NaN   
204   852936405516943360                    NaN                  NaN   
211   851953902622658560                    NaN                  NaN   
...                  ...                    ...                  ...   
784   775096608509886464                    NaN                  NaN   
794   773336787167145985                    NaN                  NaN   
800   772615324260794368                    NaN                  NaN   
811   771171053431250945                    NaN                  NaN   
815   771004394259247104                    NaN                  NaN   
818   770743923962707968                    NaN                  NaN   
822   770093767776997377                    NaN                  NaN   
826   769335591808995329                    NaN                  NaN   
829   768909767477751808                    NaN                  NaN   
833   768554158521745409                    NaN                  NaN   
841   766864461642756096                    NaN                  NaN   
847   766078092750233600                    NaN                  NaN   
860   763167063695355904                    NaN                  NaN   
868   761750502866649088                    NaN                  NaN   
872   761371037149827077                    NaN                  NaN   
885   760153949710192640                    NaN                  NaN   
890   759566828574212096                    NaN                  NaN   
895   759159934323924993                    NaN                  NaN   
908   757729163776290825                    NaN                  NaN   
911   757597904299253760                    NaN                  NaN   
926   754874841593970688                    NaN                  NaN   
937   753298634498793472                    NaN                  NaN   
943   752701944171524096                    NaN                  NaN   
949   752309394570878976                    NaN         

In [25]:
df[df.retweeted_status_id.notna()].expanded_urls.duplicated().any()

False

In [26]:
df[df.retweeted_status_id.notna()].sort_values(by='expanded_urls')

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
204   852936405516943360                    NaN                  NaN   
180   857062103051644929                    NaN                  NaN   
568   802185808107208704                    NaN                  NaN   
332   832769181346996225                    NaN                  NaN   
298   837012587749474308                    NaN                  NaN   
285   838916489579200512                    NaN                  NaN   
302   836648853927522308                    NaN                  NaN   
281   839290600511926273                    NaN                  NaN   
231   847971574464610304                    NaN                  NaN   
605   798576900688019456                    NaN                  NaN   
1043  743835915802583040                    NaN                  NaN   
273   840728873075638272                    NaN                  NaN   
604   798585098161549313                    NaN                  NaN   
603   798628517273620480                    NaN                  NaN   
1023  746521445350707200                    NaN                  NaN   
74    878316110768087041                    NaN                  NaN   
895   759159934323924993                    NaN                  NaN   
602   798644042770751489                    NaN                  NaN   
601   798665375516884993                    NaN                  NaN   
860   763167063695355904                    NaN                  NaN   
811   771171053431250945                    NaN                  NaN   
720   783347506784731136                    NaN                  NaN   
949   752309394570878976                    NaN                  NaN   
600   798673117451325440                    NaN                  NaN   
599   798682547630837760                    NaN                  NaN   
908   757729163776290825                    NaN                  NaN   
926   754874841593970688                    NaN                  NaN   
749   779124354206535695                    NaN                  NaN   
770   776819012571455488                    NaN                  NaN   
506   812747805718642688                    NaN                  NaN   
...                  ...                    ...                  ...   
78    877611172832227328                    NaN                  NaN   
327   833732339549220864                    NaN                  NaN   
160   860924035999428608                    NaN                  NaN   
1242  711998809858043904                    NaN                  NaN   
656   791780927877898241                    NaN                  NaN   
450   818646164899774465                    NaN                  NaN   
310   835309094223372289                    NaN                  NaN   
406   823719002937630720                    NaN                  NaN   
546   805823200554876929                    NaN                  NaN   
692   787111942498508800                    NaN                  NaN   
654   791821351946420224                    NaN                  NaN   
753   778774459159379968                    NaN                  NaN   
682   788552643979468800                    NaN                  NaN   
826   769335591808995329                    NaN                  NaN   
943   752701944171524096                    NaN                  NaN   
655   791784077045166082                    NaN                  NaN   
885   760153949710192640                    NaN                  NaN   
272   840761248237133825                    NaN                  NaN   
124   868639477480148993                    NaN                  NaN   
73    878404777348136964                    NaN                  NaN   
159   860981674716409858                    NaN                  NaN   
97    873337748698140672                    NaN                  NaN   
146   863471782782697472                    NaN                  NaN   
309   835536468978302976                    NaN         

In [27]:
df.query('name == "None"').name.count()
#several missing dog names

745

In [28]:
df.query('rating_numerator < 10')
#883482846933004288 should be 13.5/10
#sampling of other <10 ratings suggests they are legit

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
45    883482846933004288                    NaN                  NaN   
229   848212111729840128                    NaN                  NaN   
315   835152434251116546                    NaN                  NaN   
387   826598799820865537           8.265984e+17         4.196984e+09   
462   817502432452313088                    NaN                  NaN   
485   814578408554463233                    NaN                  NaN   
599   798682547630837760                    NaN                  NaN   
605   798576900688019456                    NaN                  NaN   
730   781661882474196992                    NaN                  NaN   
745   780092040432480260                    NaN                  NaN   
764   777953400541634568                    NaN                  NaN   
765   777885040357281792                    NaN                  NaN   
784   775096608509886464                    NaN                  NaN   
814   771014301343748096                    NaN                  NaN   
835   768193404517830656                    NaN                  NaN   
845   766423258543644672                    NaN                  NaN   
859   763183847194451968                    NaN                  NaN   
860   763167063695355904                    NaN                  NaN   
883   760252756032651264                    NaN                  NaN   
896   759099523532779520                    NaN                  NaN   
906   758041019896193024                    NaN                  NaN   
912   757596066325864448                    NaN                  NaN   
915   757354760399941633                    NaN                  NaN   
936   753375668877008896                    NaN                  NaN   
946   752568224206688256                    NaN                  NaN   
948   752334515931054080                    NaN                  NaN   
956   751583847268179968                    NaN                  NaN   
964   750506206503038976                    NaN                  NaN   
966   750383411068534784                    NaN                  NaN   
987   749036806121881602                    NaN                  NaN   
...                  ...                    ...                  ...   
2312  666776908487630848                    NaN                  NaN   
2314  666701168228331520                    NaN                  NaN   
2315  666691418707132416                    NaN                  NaN   
2316  666649482315059201                    NaN                  NaN   
2317  666644823164719104                    NaN                  NaN   
2319  666447344410484738                    NaN                  NaN   
2320  666437273139982337                    NaN                  NaN   
2322  666430724426358785                    NaN                  NaN   
2323  666428276349472768                    NaN                  NaN   
2326  666411507551481857                    NaN                  NaN   
2327  666407126856765440                    NaN                  NaN   
2328  666396247373291520                    NaN                  NaN   
2330  666362758909284353                    NaN                  NaN   
2331  666353288456101888                    NaN                  NaN   
2333  666337882303524864                    NaN                  NaN   
2334  666293911632134144                    NaN                  NaN   
2335  666287406224695296                    NaN                  NaN   
2338  666104133288665088                    NaN                  NaN   
2340  666099513787052032                    NaN                  NaN   
2341  666094000022159362                    NaN                  NaN   
2342  666082916733198337                    NaN                  NaN   
2344  666071193221509120                    NaN                  NaN   
2346  666058600524156928                    NaN                  NaN   
2347  666057090499244032                    NaN         

In [29]:
df.query('rating_numerator == 0')
#these two tweets are not valid dog ratings and can be removed

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
315   835152434251116546                    NaN                  NaN   
1016  746906459439529985           7.468859e+17         4.196984e+09   

                      timestamp  \
315   2017-02-24 15:40:31 +0000   
1016  2016-06-26 03:22:31 +0000   

                                                 source  \
315   <a href="http://twitter.com/download/iphone" r...   
1016  <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
315   When you're so blinded by your systematic plag...                  NaN   
1016  PUPDATE: can't see any. Even if I could, I cou...                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
315                        NaN                        NaN   
1016                       NaN                        NaN   

                                          expanded_urls  rating_numerator  \
315   https://twitter.com/dog_rates/status/835152434...                 0   
1016  https://twitter.com/dog_rates/status/746906459...                 0   

      rating_denominator  name doggo floofer pupper puppo  
315                   10  None  None    None   None  None  
1016                  10  None  None    None   None  None

In [30]:
df.expanded_urls.isna().sum()

59

In [31]:
df[df.expanded_urls.isna()]

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
30    886267009285017600           8.862664e+17         2.281182e+09   
55    881633300179243008           8.816070e+17         4.738443e+07   
64    879674319642796034           8.795538e+17         3.105441e+09   
113   870726314365509632           8.707262e+17         1.648776e+07   
148   863427515083354112           8.634256e+17         7.759620e+07   
179   857214891891077121           8.571567e+17         1.806710e+08   
185   856330835276025856                    NaN                  NaN   
186   856288084350160898           8.562860e+17         2.792810e+08   
188   855862651834028034           8.558616e+17         1.943518e+08   
189   855860136149123072           8.558585e+17         1.361572e+07   
218   850333567704068097           8.503288e+17         2.195506e+07   
228   848213670039564288           8.482121e+17         4.196984e+09   
234   847617282490613760           8.476062e+17         4.196984e+09   
274   840698636975636481           8.406983e+17         8.405479e+17   
290   838150277551247360           8.381455e+17         2.195506e+07   
291   838085839343206401           8.380855e+17         2.894131e+09   
313   835246439529840640           8.352460e+17         2.625958e+07   
342   832088576586297345           8.320875e+17         3.058208e+07   
346   831926988323639298           8.319030e+17         2.068372e+07   
375   828361771580813312                    NaN                  NaN   
387   826598799820865537           8.265984e+17         4.196984e+09   
409   823333489516937216           8.233264e+17         1.582854e+09   
427   821153421864615936           8.211526e+17         1.132119e+08   
498   813130366689148928           8.131273e+17         4.196984e+09   
513   811647686436880384           8.116272e+17         4.196984e+09   
570   801854953262350336           8.018543e+17         1.185634e+07   
576   800859414831898624           8.008580e+17         2.918590e+08   
611   797165961484890113           7.971238e+17         2.916630e+07   
701   786051337297522688           7.727430e+17         7.305050e+17   
707   785515384317313025                    NaN                  NaN   
843   766714921925144576           7.667118e+17         4.196984e+09   
857   763956972077010945           7.638652e+17         1.584641e+07   
967   750381685133418496           7.501805e+17         4.717297e+09   
1005  747651430853525504           7.476487e+17         4.196984e+09   
1080  738891149612572673           7.384119e+17         3.589728e+08   
1295  707983188426153984           7.079801e+17         2.319108e+09   
1345  704491224099647488           7.044857e+17         2.878549e+07   
1445  696518437233913856                    NaN                  NaN   
1446  696490539101908992           6.964887e+17         4.196984e+09   
1474  693644216740769793           6.936422e+17         4.196984e+09   
1479  693582294167244802           6.935722e+17         1.198989e+09   
1497  692423280028966913           6.924173e+17         4.196984e+09   
1523  690607260360429569           6.903413e+17         4.670367e+08   
1598  686035780142297088           6.860340e+17         4.196984e+09   
1605  685681090388975616           6.855479e+17         4.196984e+09   
1618  684969860808454144           6.849598e+17         4.196984e+09   
1663  682808988178739200           6.827884e+17         4.196984e+09   
1689  681340665377193984           6.813394e+17         4.196984e+09   
1774  678023323247357953           6.780211e+17         4.196984e+09   
1819  676590572941893632           6.765883e+17         4.196984e+09   
1844  675849018447167488           6.758457e+17         4.196984e+09   
1895  674742531037511680           6.747400e+17         4.196984e+09   
1905  674606911342424069           6.744689e+17         4.196984e+09   
1914  674330906434379776           6.658147e+17         1.637468e+07   
1940  673716320723169284           6.737159e+17         

In [32]:
df.query('tweet_id == 0').tweet_id.count()

0

### Assess: Twitter Archive Enhanced ###
**Data Quality**
* dog stage is missing for some rows which are populated with 'None' instead of null (i.e. NaN).  
* At least 62 instances of incorrect dog names (e.g. 'a' or 'an').  These words are all the 3rd word in the tweet as it appears the names were extracted with the assumption all tweets followed the pattern "This is dog_name...".
* several instances (745) of missing dog names which are populated with 'None' instead of null (i.e. NaN).  Quick scan of a few entries show these tweets are not of dogs (but of owners etc.)
* several instances (59) where the expanded url is missing
* timestamp and retweet_timestamp should be of date_time dataype
* can remove tweets that are more recent than 2017 (as we won't find predictions for those instances)
* there are 181 retweets which can be removed for our purposes (note these instances may overlap with some of the other observed issues)
* 23 instances where rating denominator is not 10 (note these instances may overlap with some of the other observed issues)
* some instances (24) where the rating_numerator is extremely high (>20) (note these instances may overlap with some of the other observed issues)
* inconsistent format for dog names (some start w/ capitals others don't)
* 70 instances of reply tweets to original do posting (additional 7 tweets where reply includes a dog and rating so can keep those)
* 2 instances where rating numerator is zero (valid issue?)

**Non-issues?**
* not clear if the different id columns (e.g. tweet_id, retweet_user_id etc.) should be numbers (e.g. int64 or float 64) as opposed to a string (leading zeros could be stripped?)
* several instances (440) where rating numerator is less than 10 (valid issue?)

**Data Tidiness**
* dog stages should probably be a single column instead of multiple columns
* new dog stages column should be of category datatype


### Gather: Image Predictions ###
**Let's gather image predictions data now**

In [33]:
folder_name = 'image_predictions'
if not os.path.exists(folder_name):
    os.makedirs(folder_name)

In [34]:
url = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'

response = requests.get(url)
with open('image_preds', mode='wb') as file:
    file.write(response.content)

In [35]:
df_p = pd.read_csv('image_preds', sep='\t')
df_p.head()

tweet_id                                          jpg_url  \
0  666020888022790149  https://pbs.twimg.com/media/CT4udn0WwAA0aMy.jpg   
1  666029285002620928  https://pbs.twimg.com/media/CT42GRgUYAA5iDo.jpg   
2  666033412701032449  https://pbs.twimg.com/media/CT4521TWwAEvMyu.jpg   
3  666044226329800704  https://pbs.twimg.com/media/CT5Dr8HUEAA-lEu.jpg   
4  666049248165822465  https://pbs.twimg.com/media/CT5IQmsXIAAKY4A.jpg   

   img_num                      p1   p1_conf  p1_dog                  p2  \
0        1  Welsh_springer_spaniel  0.465074    True              collie   
1        1                 redbone  0.506826    True  miniature_pinscher   
2        1         German_shepherd  0.596461    True            malinois   
3        1     Rhodesian_ridgeback  0.408143    True             redbone   
4        1      miniature_pinscher  0.560311    True          Rottweiler   

    p2_conf  p2_dog                   p3   p3_conf  p3_dog  
0  0.156665    True    Shetland_sheepdog  0.061428    True  
1  0.074192    True  Rhodesian_ridgeback  0.072010    True  
2  0.138584    True           bloodhound  0.116197    True  
3  0.360687    True   miniature_pinscher  0.222752    True  
4  0.243682    True             Doberman  0.154629    True

In [36]:
df_p.tail()

tweet_id                                          jpg_url  \
2070  891327558926688256  https://pbs.twimg.com/media/DF6hr6BUMAAzZgT.jpg   
2071  891689557279858688  https://pbs.twimg.com/media/DF_q7IAWsAEuuN8.jpg   
2072  891815181378084864  https://pbs.twimg.com/media/DGBdLU1WsAANxJ9.jpg   
2073  892177421306343426  https://pbs.twimg.com/media/DGGmoV4XsAAUL6n.jpg   
2074  892420643555336193  https://pbs.twimg.com/media/DGKD1-bXoAAIAUK.jpg   

      img_num           p1   p1_conf  p1_dog                  p2   p2_conf  \
2070        2       basset  0.555712    True    English_springer  0.225770   
2071        1  paper_towel  0.170278   False  Labrador_retriever  0.168086   
2072        1    Chihuahua  0.716012    True            malamute  0.078253   
2073        1    Chihuahua  0.323581    True            Pekinese  0.090647   
2074        1       orange  0.097049   False               bagel  0.085851   

      p2_dog                           p3   p3_conf  p3_dog  
2070    True  German_short-haired_pointer  0.175219    True  
2071    True                      spatula  0.040836   False  
2072    True                       kelpie  0.031379    True  
2073    True                     papillon  0.068957    True  
2074   False                       banana  0.076110   False

In [37]:
df_p.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
tweet_id    2075 non-null int64
jpg_url     2075 non-null object
img_num     2075 non-null int64
p1          2075 non-null object
p1_conf     2075 non-null float64
p1_dog      2075 non-null bool
p2          2075 non-null object
p2_conf     2075 non-null float64
p2_dog      2075 non-null bool
p3          2075 non-null object
p3_conf     2075 non-null float64
p3_dog      2075 non-null bool
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [38]:
df_p.query('img_num <= 0')

Empty DataFrame
Columns: [tweet_id, jpg_url, img_num, p1, p1_conf, p1_dog, p2, p2_conf, p2_dog, p3, p3_conf, p3_dog]
Index: []

In [39]:
df_p.query('jpg_url == "None"')

Empty DataFrame
Columns: [tweet_id, jpg_url, img_num, p1, p1_conf, p1_dog, p2, p2_conf, p2_dog, p3, p3_conf, p3_dog]
Index: []

In [40]:
df_p.tweet_id.duplicated().any()

False

In [41]:
df_p.jpg_url.duplicated().any()

True

In [42]:
df_p.jpg_url.duplicated().sum()

66

In [43]:
df_p[df_p.jpg_url.duplicated(keep=False)].sort_values(by='jpg_url').head(20)
#these appear to be duplicated images due to retweets

tweet_id                                            jpg_url  \
480   675354435921575936  https://pbs.twimg.com/ext_tw_video_thumb/67535...   
1297  752309394570878976  https://pbs.twimg.com/ext_tw_video_thumb/67535...   
1864  842892208864923648  https://pbs.twimg.com/ext_tw_video_thumb/80710...   
1641  807106840509214720  https://pbs.twimg.com/ext_tw_video_thumb/80710...   
1703  817181837579653120  https://pbs.twimg.com/ext_tw_video_thumb/81596...   
1691  815966073409433600  https://pbs.twimg.com/ext_tw_video_thumb/81596...   
1705  817423860136083457  https://pbs.twimg.com/ext_tw_video_thumb/81742...   
1858  841833993020538882  https://pbs.twimg.com/ext_tw_video_thumb/81742...   
1715  819004803107983360    https://pbs.twimg.com/media/C12whDoVEAALRxa.jpg   
1718  819015337530290176    https://pbs.twimg.com/media/C12whDoVEAALRxa.jpg   
1716  819006400881917954    https://pbs.twimg.com/media/C12x-JTVIAAzdfl.jpg   
1717  819015331746349057    https://pbs.twimg.com/media/C12x-JTVIAAzdfl.jpg   
1746  823269594223824897    https://pbs.twimg.com/media/C2kzTGxWEAEOpPL.jpg   
1738  822244816520155136    https://pbs.twimg.com/media/C2kzTGxWEAEOpPL.jpg   
1740  822489057087389700    https://pbs.twimg.com/media/C2oRbOuWEAAbVSl.jpg   
1742  822647212903690241    https://pbs.twimg.com/media/C2oRbOuWEAAbVSl.jpg   
1767  826958653328592898    https://pbs.twimg.com/media/C3nygbBWQAAjwcW.jpg   
1789  829878982036299777    https://pbs.twimg.com/media/C3nygbBWQAAjwcW.jpg   
1785  829374341691346946    https://pbs.twimg.com/media/C4KHj-nWQAA3poV.jpg   
1903  851953902622658560    https://pbs.twimg.com/media/C4KHj-nWQAA3poV.jpg   

      img_num                         p1   p1_conf  p1_dog  \
480         1                    upright  0.303415   False   
1297        1                    upright  0.303415   False   
1864        1                  Chihuahua  0.505370    True   
1641        1                  Chihuahua  0.505370    True   
1703        1            Tibetan_mastiff  0.506312    True   
1691        1            Tibetan_mastiff  0.506312    True   
1705        1                   ice_bear  0.336200   False   
1858        1                   ice_bear  0.336200   False   
1715        1            standard_poodle  0.351308    True   
1718        1            standard_poodle  0.351308    True   
1716        4                     prison  0.907083   False   
1717        4                     prison  0.907083   False   
1746        1                    Samoyed  0.585441    True   
1738        1                    Samoyed  0.585441    True   
1740        1                    Samoyed  0.416769    True   
1742        1                    Samoyed  0.416769    True   
1767        1           golden_retriever  0.617389    True   
1789        1           golden_retriever  0.617389    True   
1785        1  Staffordshire_bullterrier  0.757547    True   
1903        1  Staffordshire_bullterrier  0.757547    True   

                                  p2   p2_conf  p2_dog  \
480                 golden_retriever  0.181351    True   
1297                golden_retriever  0.181351    True   
1864                      Pomeranian  0.120358    True   
1641                      Pomeranian  0.120358    True   
1703                 Tibetan_terrier  0.295690    True   
1691                 Tibetan_terrier  0.295690    True   
1705                         Samoyed  0.201358    True   
1858                         Samoyed  0.201358    True   
1715                      toy_poodle  0.271929    True   
1718                      toy_poodle  0.271929    True   
1716                          palace  0.020089   False   
1717                          palace  0.020089   False   
1746                      Pomeranian  0.193654    True   
1738                      Pomeranian  0.193654    True   
1740                        malamute  0.252706    True   
1742                        malamute  0.252706    True   
1767              Labrador_retriever  0.337053    True   
1789     

In [44]:
df_p.p1_dog.mean(), df_p.p1_dog.mean()

(0.7383132530120482, 0.7383132530120482)

In [45]:
df_p.p2_dog.mean(), df_p.p2_dog.mean()
# it appears both the prediction confidence and accuracy is slightly better in prediction 2, however only a minor difference b/w the 3 predictions


(0.74843373493975907, 0.74843373493975907)

In [46]:
df_p.p3_dog.mean(), df_p.p3_dog.mean()

(0.72240963855421692, 0.72240963855421692)

### Assess: Image Predictions ###
**Quality**
* There are 2075 predictions, whereas there are 2356 tweets meaning there are tweet records that either  do not have images or are more recent than 2017 or are retweet duplicates -- in either case, these instances can be removed
* There are 66 instances of duplicate predications whereby both the original tweet and the retweeted image were used

**Tidiness**
* Each row comprises 3 observations (i.e. 3 predictions).  If desired, instead of 9 separate columns for each of 3 dog predictions and associated confidence and accuracy, can collapse into three columns: one column indicating prediction #, another column indicating prediction confidence and accuracy

### Gather: Twitter API ###
**Now we'll gather additional data from the Twitter API**
* retweet count 
* favorite count

In [47]:
#import tweepy -- already did this earlier
#actual keys/secrets/tokens will need to be obfuscated
#code to create an API object that you can use to gather Twitter data.
consumer_key = 'MY CONSUMER KEY'
consumer_secret = 'MY CONSUMER SECRET'
access_token = 'MY ACCESS TOKEN'
access_secret = 'MY ACCESS SECRET'

auth = tweepy.OAuthHandler(consumer_key, consumer_secret)
auth.set_access_token(access_token, access_secret)

api = tweepy.API(auth, wait_on_rate_limit = True, wait_on_rate_limit_notify = True)

In [48]:
# import json
# import time

# start_time = time.time()
# loop_count = 0

# with open('tweet_json.txt','w', encoding = 'utf8') as outfile:
#     for tweet_id in df.tweet_id:
#         loop_count += 1
#         try:
#             tweet_extend = api.get_status(tweet_id, tweet_mode='extended') # set mode to extended
#             json.dump(tweet_extend._json, outfile)
#             outfile.write('\n') # add a line to separate different json files
#             if loop_count % 100 == 0: # to minimize number of messages, let's only count every 100 loops
#                 print("loop # " + str(loop_count))                
#                 print('trying tweet_id {} now'.format(tweet_id))
#                 elapsed_time = time.time() - start_time
#                 elapsed_time_str = time.strftime("%H:%M:%S", time.gmtime(elapsed_time))
#                 print("\nTime elapsed (HH:MM:SS) so far:  " + elapsed_time_str + "\n")            
#         except Exception as e: 
#             print('for tweet_id {} the following error occured: {} '.format(tweet_id, e))
#         pass

# elapsed_time = time.time() - start_time
# elapsed_time_str = time.strftime("%H:%M:%S", time.gmtime(elapsed_time))
# print("\nTotal time elapsed (HH:MM:SS):  " + elapsed_time_str + "\n")

In [49]:
import json
tweets = []

tweet_json = open('tweet_json.txt', 'r')

for line in tweet_json:
    tweet = json.loads(line)
    tweets.append(tweet)

In [50]:
import json

tweets = [] #empty list for storing tweet info to be converted to Dataframe after population

with open('tweet_json.txt', 'r') as json_file:
    #data = json.loads(json_file)
    for line in json_file:
        tweet = json.loads(line)
        json_data = {'tweet_id': tweet['id'],  #create dictionary for each tweet's desired fields
        'retweet_count': tweet['retweet_count'], 
        'favorite_count': tweet['favorite_count']}
        tweets.append(json_data) #append fields to list
#after reading all tweets, convert list of dictionaries to a dataframe
df_tweets = pd.DataFrame(tweets, columns = ['tweet_id',
                                              'retweet_count',
                                              'favorite_count'])
#inspect dataframe
df_tweets.head()

tweet_id  retweet_count  favorite_count
0  892420643555336193           7966           37191
1  892177421306343426           5913           31993
2  891815181378084864           3909           24092
3  891689557279858688           8115           40511
4  891327558926688256           8799           38741

In [51]:
df_tweets.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2332 entries, 0 to 2331
Data columns (total 3 columns):
tweet_id          2332 non-null int64
retweet_count     2332 non-null int64
favorite_count    2332 non-null int64
dtypes: int64(3)
memory usage: 54.7 KB


In [52]:
df.tweet_id.count() - df_tweets.tweet_id.count()

24

### Assess: Twitter API Data ###

**Quality**
* There were 24 instances where the twitter API encountered an error retrieving data.  May want to exclude those records when analyzing in combination with twitter archive data

**Tidiness**
* The Twitter API data can be combined with the twitter archive into a single dataframe for easier analysis

In [53]:
with open('tweet_json.txt', 'r') as json_file:
    #data = json.loads(json_file)
    for line in json_file:
        tweet = json.loads(line)
        json_data = {'tweet_id': tweet['id'],  #create dictionary for each tweet's desired fields
        'retweet_count': tweet['retweet_count'], 
        'favorite_count': tweet['favorite_count']}
        tweets.append(json_data) #append fields to list
#after reading all tweets, convert list of dictionaries to a dataframe
df_tweets = pd.DataFrame(tweets, columns = ['tweet_id',
                                              'retweet_count',
                                              'favorite_count'])

## Define ##
**Let's define the fixes we want to implement for the different data quality and tidiness issues**


### Assess: Twitter Archive Enhanced ###
**Data Quality**
* Replace instances of 'None' with NaN for dog stage.
* Replace the 745 instances of 'None' with NaN for missing dog names.
* change timestamp and retweet_timestamp to date_time dataype
* remove tweets that are more recent than 2017 (as we won't find predictions for those instances)
* remove the 181 retweets
* Manually correct the roughly 20 extremely high ratings (>20) --> these appear to have extracted the rating incorrectly due to inconsistent tweet text string patterns.
* Manually correct the roughly 20 instances where the rating denominator is not equal to 20 --> these appear to have extracted the rating incorrectly due to inconsistent tweet text string patterns.
* Remove the roughly 70 instances of reply tweets to original posting (additional 7 tweets where reply includes a dog and rating so can keep those)
* Can remove the 2 instances of ratings of zero (0) as those are not for actual dog postings (tweets are about plagiarism and a poor view from the empire state building)

* (observation) At least 62 instances of incorrect dog names (e.g. 'a' or 'an').  These words are all the 3rd word in the tweet as it appears the names were extracted with the assumption all tweets followed the pattern "This is dog_name...": We don't really have a programmable way to fix this yet so will leave outstanding for now: 
* (observation) several instances (59) where the expanded url is missing: If desired, we can create the expanded url using the tweet_id concatenated to "https://twitter.com/dog_rates/status/"
* (Observation) inconsistent format for dog names (some start w/ capitals others don't): I don't think this will affect our analysis so we may leave as is.

**Non-issues?**
* not clear if the different id columns (e.g. tweet_id, retweet_user_id etc.) should be numbers (e.g. int64 or float 64) as opposed to a string (leading zeros could be stripped?)
* several instances (440) where rating numerator is less than 10 (valid issue?)
* 2 instances where rating numerator is zero (valid issue?)

**Data Tidiness**
* Use Melt function to change the multiple columns for dog stages into a single column and assign datatype of category to new column


### Assess: Image Predictions ###
**Quality**
* Remove the 66 instances of duplicate predictions (i.e. whereby both the original tweet and the retweeted image were used)
* There are 2075 predictions, whereas there are 2356 tweets in the archive meaning there are tweet records that either  do not have images or are more recent than 2017 or are retweet duplicates: for now, let's keep these as we may want to use those values for other analysis

**Tidiness**
* Each row comprises 3 observations (i.e. 3 predictions).  If desired, instead of 9 separate columns for each of 3 dog predictions and associated confidence and accuracy, can collapse into three columns: one column indicating prediction #, another column indicating prediction confidence and accuracy: not yet clear if this needs to be cleaned up, but if so we could use the Melt function again here.

### Assess: Twitter API Data ###

**Quality**
* After combining with the twitter archive data (see tidiness fix), we'll remove from the original twitter archive the 24 instances where the twitter API encountered an error retrieving data.

**Tidiness**
* The Twitter API data can be combined with the twitter archive into a single dataframe for easier analysis